engine/utils/HeapArray.h

```diff
@@ -0,0 +1,735 @@
+//
+//  Copyright (c) 2022 Jonathan Moallem (@J-Mo63) & Aryeh Zinn (@Raelr)
+//
+//  This code is released under an unmodified zlib license.
+//  For conditions of distribution and use, please see:
+//      https://opensource.org/licenses/Zlib
+//
+
+#ifndef SIEGE_ENGINE_HEAP_ARRAY_H
+#define SIEGE_ENGINE_HEAP_ARRAY_H
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <initializer_list>
+#include <utility>
+
+typedef const size_t sizet;
+
+namespace Siege::Utils
+{
+
+// Static methods
+
+/**
+ * @brief Allocates a chunk of memory and returns it as a pointer to a type.
+ * @tparam T the type of the pointer to allocate.
+ * @param dataSize the size of the memory we want to allocate.
+ * @return the pointer we allocated.
+ */
+template<typename T>
+static inline T* Allocate(const size_t& dataSize)
+{
+    return static_cast<T*>(malloc(dataSize));
+}
+
+/**
+ * @brief Reallocate a chunk of memory to a new size.
+ * @tparam T the type of the pointer.
+ * @param ptr the pointer to be reallocated.
+ * @param dataSize the size of the new memory chunk.
+ * @return the new reallocated pointer.
+ */
+template<typename T>
+static inline T* Reallocate(T* ptr, const size_t& dataSize)
+{
+    return static_cast<T*>(realloc(ptr, dataSize));
+}
```

Comment:
These functions should be moved to the source file

Reply:
I don't understand what you mean by this. The fact that these are templated means that they can't be moved to a `.cpp` file. They have to live in the header.

Reply:
Remember to move this to another namespace so it doesn't collide